### DISCUSSION TITLE
Using large voxel numbers in Dosxyznrc

### DISCUSSION THREAD

**Discussion by Abdullah-Abuhaimed** [nrc-cnrc/EGSnrc#723] [Questions] [answered]
Dear all,

I am running simulations in Dosxyznrc using an obese full body phantom that has a large number of voxels (x=350, y=410, and z=652). When I change (IMAX,JMAX, and KMAX) in dosxyznrc_user_macros.mortran to these values, I get the error shown below. I have managed to run simulations with smaller phantoms with no issue, but not for large phantoms. It would be greatly appreciated if anyone could assist me in overcoming this issue.

```make 
Fortran compiling dosxyznrc_linux.F using flags ' -fPIC -O2 -mtune=native' and extra objects/libs '/project/k1508/EGSnrc/HEN_HOUSE/lib/linux/egs_c_utils.o /project/k1508/EGSnrc/HEN_HOUSE/lib/linux/read_write_pardose.o /project/k1508/EGSnrc/HEN_HOUSE/lib/linux/load_beamlib.o  /project/k1508/EGSnrc/HEN_HOUSE/lib/linux/load_vculib.o  -ldl -L/project/k1508/EGSnrc/HEN_HOUSE/egs++/dso/linux -Wl,-rpath,/project/k1508/EGSnrc/HEN_HOUSE/egs++/dso/linux -liaea_phsp'
/usr/bin/ld: failed to convert GOTPCREL relocation; relink with --no-relax
/usr/bin/sha1sum: /project/k1508/EGSnrc/egs_home/bin/linux/dosxyznrc: No such file or directory
collect2: error: ld returned 1 exit status
make: *** [/project/k1508/EGSnrc/HEN_HOUSE/makefiles/standard_makefile:90: /project/k1508/EGSnrc/egs_home/bin/linux/dosxyznrc] Error 1
```

**Comment by rtownson** [comment, marked as answer]
Try changing the compiler flags in the Makefile to allow for large arrays:
`EGS_EXTRA_LIBS = $(FLIBS) $(BEAMLIB_EXTRA_LIBS) -D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE -mcmodel=large`

This is a common question, there's more discussion here: https://www.reddit.com/r/EGSnrc/comments/jkcv4e/increase_the_maximun_number_of_voxels_in_xyz/

**Reply by crcrewso** [reply]
I'm wondering, with GCC is it time we add this to the default flags?

> -mcmodel=medium
>Generate code for the medium model: the program is linked in the lower 2 GB of the address space. Small symbols are also placed there. Symbols with sizes larger than -mlarge-data-threshold are put into large data or BSS sections and can be located above 2GB. Programs can be statically or dynamically linked.
>-mcmodel=large
>Generate code for the large model. This model makes no assumptions about addresses and sizes of sections.
[Link](https://gcc.gnu.org/onlinedocs/gcc/x86-Options.html)

**Reply by Abdullah-Abuhaimed** [reply]
Thank you Reid for your response, 

I have tried to add the options (all together or individually), but I got the error shown below. I modified both the Makefile in HEN_HOUSE/ and-or egs_home/, but sill no luck. Any idea?

```make
Fortran compiling dosxyznrc_linux.F using flags ' -fPIC -O2 -mtune=native' and extra objects/libs '/project/k1508/EGSnrc/HEN_HOUSE/lib/linux/egs_c_utils.o /project/k1508/EGSnrc/HEN_HOUSE/lib/linux/read_write_pardose.o /project/k1508/EGSnrc/HEN_HOUSE/lib/linux/load_beamlib.o  /project/k1508/EGSnrc/HEN_HOUSE/lib/linux/load_vculib.o  -ldl -L/project/k1508/EGSnrc/HEN_HOUSE/egs++/dso/linux -Wl,-rpath,/project/k1508/EGSnrc/HEN_HOUSE/egs++/dso/linux -liaea_phsp -D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE -mcmodel=large'
/usr/bin/ld: failed to convert GOTPCREL relocation; relink with --no-relax
/usr/bin/sha1sum: /project/k1508/EGSnrc/egs_home/bin/linux/dosxyznrc: No such file or directory
collect2: error: ld returned 1 exit status
make: *** [/project/k1508/EGSnrc/HEN_HOUSE/makefiles/standard_makefile:90: /project/k1508/EGSnrc/egs_home/bin/linux/dosxyznrc] Error 1 
```

**Reply by rtownson** [reply]
You may need to re-run the installer (or HEN_HOUSE/scripts/configure if you're on linux) and add the flags for the fortran compilation there. This was discussed in the link I shared so maybe that was the extra step they took that worked. Also make sure you have enough RAM - that array size is nearly 1GB and it needs to be able to allocate several of them at once.

@crcrewso you're probably right, we should at least add -mcmodel=medium and see if that covers it.

**Reply by crcrewso** [reply]
Done, see #725

**Reply by Abdullah-Abuhaimed** [reply]
Thank you @rtownson for your support. I have managed to get the simulation worked by re-running the installer and adding the suggested options in the compiler flags, not only for fortran, but also for the other compilers.

**Comment by ftessier** [comment]
The `-mcmodel=medium` option has been added to the default compiler flags in the configure script and configuration gui. See #725.
